cpp/src/arrow/parquet/parquet-io-test.cc

```diff
@@ -0,0 +1,225 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gtest/gtest.h"
+
+#include "arrow/test-util.h"
+#include "arrow/parquet/reader.h"
+#include "arrow/parquet/writer.h"
+#include "arrow/types/primitive.h"
+#include "arrow/util/memory-pool.h"
+#include "arrow/util/status.h"
+
+#include "parquet/api/reader.h"
+#include "parquet/api/writer.h"
+
+using ParquetBuffer = parquet::Buffer;
+using parquet::BufferReader;
+using parquet::InMemoryOutputStream;
+using parquet::ParquetFileReader;
+using parquet::ParquetFileWriter;
+using parquet::RandomAccessSource;
+using parquet::Repetition;
+using parquet::SchemaDescriptor;
+using ParquetType = parquet::Type;
+using parquet::schema::GroupNode;
+using parquet::schema::NodePtr;
+using parquet::schema::PrimitiveNode;
+
+namespace arrow {
+
+namespace parquet {
+
+class TestParquetIO : public ::testing::Test {
+ public:
+  virtual void SetUp() {}
+
+  std::shared_ptr<GroupNode> Int64Schema() {
+    auto pnode = PrimitiveNode::Make("int64", Repetition::REQUIRED, ParquetType::INT64);
+    NodePtr node_ =
+        GroupNode::Make("schema", Repetition::REQUIRED, std::vector<NodePtr>({pnode}));
+    return std::static_pointer_cast<GroupNode>(node_);
+  }
+
+  std::shared_ptr<GroupNode> DoubleSchema() {
+    auto pnode = PrimitiveNode::Make("double", Repetition::OPTIONAL, ParquetType::DOUBLE);
+    NodePtr node_ =
+        GroupNode::Make("schema", Repetition::REQUIRED, std::vector<NodePtr>({pnode}));
+    return std::static_pointer_cast<GroupNode>(node_);
+  }
+
+  std::shared_ptr<PrimitiveArray> DoubleValueArray(
+      size_t size, double value, size_t num_nulls) {
+    std::vector<double> values(size, value);
+    std::vector<uint8_t> valid_bytes(size, 1);
+
+    for (size_t i = 0; i < num_nulls; i++) {
+      valid_bytes[i * 2] = 0;
+    }
+
+    DoubleBuilder builder(default_memory_pool(), std::make_shared<DoubleType>());
+    builder.Append(values.data(), values.size(), valid_bytes.data());
+    return std::static_pointer_cast<PrimitiveArray>(builder.Finish());
+  }
+
+  std::shared_ptr<PrimitiveArray> Int64ValueArray(size_t size, int64_t value) {
+    std::vector<int64_t> values(size, value);
+    Int64Builder builder(default_memory_pool(), std::make_shared<Int64Type>());
+    builder.Append(values.data(), values.size());
+    return std::static_pointer_cast<PrimitiveArray>(builder.Finish());
+  }
```

Comment:
If you aren't using any member variables, you can put this outside the test fixture. That may be able to assist, then, with creating parameterized tests (having a corresponding faux data generation function for each data type)